Styling refactor for basic capa problems

marcotuts opened the pull request:
**Description**
The following is an in progress PR to update the styling for basic capa input problems to improve both the web and mobile experiences. For multiple choice, checkboxes, dropdowns, and text / number input problems, this styling will make inputs have clearer labels, hover states, larger clickable areas, and cleaner SASS behind the scenes.  This work is in support of an effort to put assessments into the mobile application. 
 
- - -
 
**JIRA Story** ([UX-2185](https://openedx.atlassian.net/browse/UX-2185))
**Confluence / Product Asset** ([Capa Styling Updates for Mobile Assessments](https://openedx.atlassian.net/wiki/display/MA/Capa+Styling+Updates+for+Mobile+Assessments))
**Sandbox URL** (Last Updated: 06/22) [pr8151.m.sandbox.edx.org](http://pr8151.m.sandbox.edx.org/courses/course-v1:edX+34245+12345524109/courseware/02f93ffe02cc48539ebeedf282b8f17e/c7f6939944a34a62a67bca63b93e48ca/) 
**Dependencies** Testing and review of these and more complex capa problems in real courses. 
- - -
**Additional Resources**
Test Course: ([Capa Input Test Course](https://edge.edx.org/courses/course-v1:edX+101+Spring/courseware/02f93ffe02cc48539ebeedf282b8f17e/c7f6939944a34a62a67bca63b93e48ca/))
Usage Note  - make improvements to this test course on edge, then everyone can import/ export the latest version of this test course! 

Video Summarizing Changes: 
([Video Link](https://drive.google.com/file/d/0B7uurO2TaNyfS0gySUhNd0tSbUk/view?usp=sharing))
- - -
**PR Author(s) Notes / To-Do** A list of known open tasks that the PR author has.
- [x] review / test dropdown problems
- [x] review / test text input problems
- [x] review / test number input problems
- [x] review RTL 
- [x] remove 2 cherry-picked mobile commits from Frances, those are merging in via a separate PR
- [x] i18n support for data-tooltip fields added. 
- [x] review and correct issue with BodyX course
- [x] add data-tooltip to numerical input
- [x] fix any failing tests
- [x] update test course to cover all test cases

- - -
**Test Courses** 
- [x] JusticeX
- [x] MRevX
- [x] DemoX
- [x] Test new OLI hinting templates that merged last week. (now rebased to include these changes)

- - -
**Open Bugs, Review Feedback** 
- [x] @Kyle : remove icon for problem status on text input (to match checkbox / dropdown problems)
- [x] review test coverage report / quality. 
- [x] review MRev issue with non-rendering Capa buttons: Topic 1: Intro to Body101x Lesson 1: Overview: Activity 3
- [x] regression: dropdown problem types no longer showing state icons. @Marcotuts to review / fix. 
- [x] story created to improve styling for longer inputs. Not included in this PR.  (https://openedx.atlassian.net/browse/UX-2384) 
- [x] log bug around space above text and numerical inputs, which seems to be missing from time to time without clear explanation. (https://openedx.atlassian.net/browse/TNL-2553)
- [x] remove $cms-blue rebase issues!  (frrrances review)
- [x] log separate LMS bug for improving error on numerical inputs when learners use text (https://openedx.atlassian.net/browse/TNL-2552)

- - -
**Screenshots** 
Multiple Choice, multiple question problem, explanation enabled, All other settings default 
![image](https://cloud.githubusercontent.com/assets/2023680/7752821/a97bf036-ffaf-11e4-85b8-8c1cbc23650f.png)

Shows data-tooltip for "Unanswered" icon (not a font-awesome question mark instead of a circle)
![image](https://cloud.githubusercontent.com/assets/2023680/7752830/bb87f9be-ffaf-11e4-98b9-86a1360feb4c.png)

Shows the impact of styling on a mobile viewport
![image](https://cloud.githubusercontent.com/assets/2023680/7752854/d5a886a6-ffaf-11e4-8753-c0d705c6192f.png)
 
- - -
**Reviewers**
- [x] @clrux
- [x] @frrrances 

- - - 
**Product Verification**
(link to sandbox above goes directly to problems. log in as staff.) 
- [x] @lou-wang 
- [ ] @explorerleslie  

- - - 
**Participants / Contributors**
@mekkz, @marcotuts

mekkz commented:
One bug I found: When your answer is marked as correct/incorrect, and you change the value of your answer, the indicator turns back into a question mark, but the tool tip doesn't update along with it. 
![image](https://cloud.githubusercontent.com/assets/3628148/7995018/326760ca-0ae0-11e5-9648-4f2513af3ab3.png)

mekkz commented:
Whoops, looks like I screwed up a bunch of tests with that last commit. I'll fix that shortly

clrux commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -58,7 +39,23 @@
         <span id="answer_${id}"></span>
     </fieldset>
-
+    <div class="indicator_container">
```
NIT: use dashes rather than underscores for classnames.

marcotuts replied:
updated all instances of this variable to use dashes

clrux commented on common/lib/capa/capa/templates/choicetext.html:
```diff
@@ -9,12 +9,7 @@
 <section id="choicetextinput_${id}" class="choicetextinput">
 <form class="choicetextgroup capa_inputtype" id="inputtype_${id}">
     <div class="script_placeholder" data-src="${STATIC_URL}js/capa/choicetextinput.js"/>
```
NIT: use dashes instead of underscores

clrux commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -58,7 +39,23 @@
-
+    <div class="indicator_container">
+        % if input_type == 'checkbox' or not value:
+            <span class="status ${status.classname if show_correctness != 'never' else 'unanswered'}"
```
Can we get a more semantic or meaningful classname here? `status` feels very general.

clrux commented on common/lib/capa/capa/templates/choicetext.html:
```diff
@@ -62,6 +57,13 @@
     </fieldset>
         <input class= "choicetextvalue" type="hidden"  name="input_${id}{}" id="input_${id}" value="${value|h}" />
+    
+    <div class="indicator_container">
```
NIT: use dashes instead of underscores for classnames

clrux commented on common/lib/capa/capa/templates/choicetext.html:
```diff
@@ -62,6 +57,13 @@
+    
+    <div class="indicator_container">
+    % if input_type == 'checkbox' or not element_checked:
+        <span class="status ${status.classname}" id="status_${id}"></span>
```
Can we use a more meaningful classname here?

clrux commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -58,7 +39,23 @@
+        % endif
+    </div>
     % if show_correctness == "never" and (value or status not in ['unsubmitted']):
     <div class="capa_alert">${submitted_message}</div>
```
NIT: use dashes instead of underscores for classnames

clrux commented on common/lib/capa/capa/templates/choicetext.html:
```diff
@@ -62,6 +57,13 @@
         <span id="answer_${id}"></span>
     </fieldset>
         <input class= "choicetextvalue" type="hidden"  name="input_${id}{}" id="input_${id}" value="${value|h}" />
```
NIT: space between `=` and `"` in class.

clrux commented on common/lib/capa/capa/templates/choicetext.html:
```diff
@@ -62,6 +57,13 @@
+    </div>
+
     % if show_correctness == "never" and (value or status not in ['unsubmitted']):
     <div class="capa_alert">${_(submitted_message)}</div>
```
NIT: use dashes instead of underscores for classnames

clrux commented on common/lib/capa/capa/templates/formulaequationinput.html:
```diff
@@ -10,9 +10,11 @@
-      <p class="status" id="${id}_status">
-           ${status.display_name}
-      </p>
+      <span class="status" id="${id}_status" data-tooltip="${status.display_tooltip}">
```
Can we use a more meaningful classname here?

clrux commented on common/lib/capa/capa/templates/optioninput.html:
```diff
@@ -13,12 +13,13 @@
-    <span class="sr">${value|h} - ${status.display_name}</span>
-  </span>
-
+  <div class="indicator_container">
```
NIT: use dashes instead of underscores for classnames

clrux commented on common/lib/capa/capa/templates/textline.html:
```diff
@@ -27,18 +27,20 @@
    ${trailing_text | h}
 
-      <p class="status" 
+      <span class="status" 
```
Can we use a more meaningful classname here?

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -61,51 +114,100 @@ div.problem {
+      margin-bottom: ($baseline/2);
+      padding: ($baseline/2);
 
       &.choicegroup_correct {
```
NIT: use dashes instead of underscores for classnames

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -61,51 +114,100 @@ div.problem {
         }
       }
 
       &.choicegroup_incorrect {
```
NIT: use dashes instead of underscores for classnames

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -61,51 +114,100 @@ div.problem {
+      }
     }
 
     .indicator_container {
```
NIT: use dashes instead of underscores for classnames

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -61,51 +114,100 @@ div.problem {
+      @include margin(($baseline/4) ($baseline/2) ($baseline/4) ($baseline/4));
     }
 
     text {
```
Is `text` an element?

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -61,51 +114,100 @@ div.problem {
+// Summary status indicators shown after the input area 
+div.problem { 
+
+  .indicator_container {
```
NIT: use dashes instead of underscores for classnames

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -61,51 +114,100 @@ div.problem {
+      // CASE: processing
+      &.processing {
+        // add once spinner is rotated through animations
+        //@include status-icon($gray-d1, "\f110", 0);
```
Can we remove this if it's not being used?

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -151,17 +253,22 @@ div.problem {
+
       &.status {
-        margin: 8px 0 0 $baseline/2;
+        @include margin(8px, 0, 0, $baseline/2);
```
Can you wrap the last value here in another set of quotes? I., `(8px, 0, 0, ($baseline/2))`

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -548,6 +661,93 @@ div.problem {
+// +Problem - Text Input, Numerical Input
+// ==================== 
+.problem {
+  .capa_inputtype.textline, .inputtype.formulaequationinput {
```
NIT: use dashes instead of underscores for classnames

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -548,6 +661,93 @@ div.problem {
+// ==================== 
+div.problem { 
 
   .CodeMirror {
```
NIT: can we use all lowercase letters with dashes between whole words?

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -598,7 +798,51 @@ div.problem {
+  .save, .check, .show, .reset {
+    @include margin-right($baseline/2);
+    margin-bottom: ($baseline/2);
+    height: ($baseline*2);
```
NIT: property order

clrux replied:
NIT: property order

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -598,7 +798,51 @@ div.problem {
+    // margin-top: ($baseline/2);
+    display: inline-block;
+    margin-top: 8px;
+    @include margin-left($baseline/2);
```
NIT: property order

clrux commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -466,7 +466,7 @@ class @Problem
     formulaequationinput: (element) ->
       $(element).find('input').on 'input', ->
-        $p = $(element).find('p.status')
+        $p = $(element).find('span.status')
```
Can this be done without the element (using only the class)?

clrux commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -494,7 +494,7 @@ class @Problem
     textline: (element) ->
       $(element).find('input').on 'input', ->
-        $p = $(element).find('p.status')
+        $p = $(element).find('span.status')
```
Can this be done without the element (using only the class)?

clrux commented on common/test/acceptance/pages/lms/problem.py:
```diff
@@ -45,7 +45,7 @@ def is_correct(self):
         Is there a "correct" status showing?
         """
-        return self.q(css="div.problem div.capa_inputtype.textline div.correct p.status").is_present()
+        return self.q(css="div.problem div.capa_inputtype.textline div.correct span.status").is_present()
```
Can this be done without the element (using only the class)?

clrux commented on lms/templates/instructor/instructor_dashboard_2/instructor_dashboard_2.html:
```diff
@@ -1,9 +1,13 @@
 <%inherit file="/main.html" />
 <%namespace name='static' file='/static_content.html'/>
+<<<<<<< HEAD
```
Looks like some conflicts made it into this branch. Please clean this up.

clrux commented on lms/templates/instructor/instructor_dashboard_2/instructor_dashboard_2.html:
```diff
@@ -1,9 +1,13 @@
 from django.utils.translation import ugettext as _
 from django.core.urlresolvers import reverse
 %>
+=======
```
Looks like some conflicts made it into this branch. Please clean this up.

clrux commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -58,7 +39,23 @@
+    <div class="indicator_container">
+        % if input_type == 'checkbox' or not value:
+            <span class="status ${status.classname if show_correctness != 'never' else 'unanswered'}"
+                id="status_${id}"
```
Are there line breaks here due to a max line length, or for visual formatting? I'd prefer it if these were on one line if possible.

clrux commented on common/lib/capa/capa/templates/optioninput.html:
```diff
@@ -13,12 +13,13 @@
-  </span>
-
+  <div class="indicator_container">
+    <span class="status ${status.classname}"
```
Are there line breaks here due to a max line length, or for visual formatting? I'd prefer it if these were on one line if possible.

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -61,51 +114,100 @@ div.problem {
+      width: 100%;
+      border: 2px solid $gray-l4;
+      border-radius: 3px;
+      margin-bottom: ($baseline/2);
```
NIT: property order

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -61,51 +114,100 @@ div.problem {
+      border: 2px solid $gray-l4;
+      border-radius: 3px;
+      margin-bottom: ($baseline/2);
+      padding: ($baseline/2);
```
NIT: property order

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -61,51 +114,100 @@ div.problem {
+  .indicator_container {
+
+    .status {
+      width: 20px;
```
Use the `$baseline` variable for both of these.

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -205,7 +312,21 @@ div.problem {
+      }
+
+      input {
+        border-color: red;
```
Please use the `$red` variable or similar from `lms/static/sass/base_variables.scss`.

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -205,7 +312,21 @@ div.problem {
+
       p.status {
         display: inline-block;
         width: 20px;
```
Use `$baseline` variable.

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -548,6 +661,93 @@ div.problem {
+      border: 2px solid $gray-l4;
+      border-radius: 3px;
+      height: 46px;
+      min-width: 160px;
```
NIT: property order

clrux commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -548,6 +661,93 @@ div.problem {
+      .status {
+        display: inline-block;
+        background: none;
+        margin-top: ($baseline/2);
```
NIT: property order

clrux commented on common/static/sass/_mixins.scss:
```diff
@@ -428,3 +429,11 @@
+// * +Icon - Font-Awesome  - Extend
+// ==================== 
+%use-font-awesome {
+  font-family: FontAwesome;
```
NIT: property order

clrux commented on lms/static/sass/base/_layouts.scss:
```diff
@@ -0,0 +1,104 @@
+
+    .nav-wrapper {
+      min-width: auto;
+      padding-right: 2%;
```
Are the percentages for mobile/responsive? If not, can we use fixed variables for now? You may also use `flex-grid` which is offered by Bourbon/Neat.

clrux commented on lms/static/sass/base/_layouts.scss:
```diff
@@ -0,0 +1,104 @@
+
+  .wrapper-course-material .course-material,
+  .wrapper-preview-menu .preview-menu  {
+    padding: 15px 2%;
```
Are the percentages for mobile/responsive? If not, can we use fixed variables for now? You may also use `flex-grid` which is offered by Bourbon/Neat.

clrux commented on lms/static/sass/base/_layouts.scss:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  .container {
+    max-width: none;
```
Are these `max` and `min` values overrides? Because they are being overridden by the following `width` value.

clrux commented on lms/static/sass/base/_layouts.scss:
```diff
@@ -0,0 +1,104 @@
+  .course-license {
+    max-width: 1180px;
+    margin: 0 auto;
+    padding-right: 2%;
```
Are the percentages for mobile/responsive? If not, can we use fixed variables for now? You may also use flex-grid which is offered by Bourbon/Neat.

clrux commented on lms/static/sass/base/_layouts.scss:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  .wrapper-footer {
+    padding-right: 2%;
```
Are the percentages for mobile/responsive? If not, can we use fixed variables for now? You may also use flex-grid which is offered by Bourbon/Neat.

clrux commented:
@marcotuts I've gone through and done a review of the FED here. Visually, when running this branch on the Body101x class, I see the following:

![screen shot 2015-06-08 at 1 58 12 pm](https://cloud.githubusercontent.com/assets/2112024/8041660/ebbf666c-0de7-11e5-9f1f-d1e1f77d821d.png)

It looks like the positioning of some of the "?"s are off and/or behind form fields.

mekkz commented:
Jenkins test this please

mekkz commented:
Jenkins test this please (again...)

marcotuts commented:
@clrux I pushed a commit with some cleanup based on your feedback. I had a few notes though before another review:

1.) I didn't address feedback from the second commit (cherry-picked @frrrances's commit into this PR), since that is being reviewed in a different PR and I can't comment on some of the fixes directly. 

2.) I didn't change all the NITs on class naming to use dashes for styles I didn't modify, in large part because I was concerned that it might change the scope of review / testing to be much larger given all the wonkiness of capa. Would love to address these things in a larger capa cleanup though. 

3.) Thanks for noting the BodyX course issue, we can review that case. It isn't fixed yet but once it is I can ping you for another review. Thanks!

marcotuts commented:
I reviewed the BodyX course and found that the styling changes are due to some more heavy handed table styling used by the BodyX team for their capa problems. (cc'ed @clrux, @explorerleslie) I believe that this isn't a bug, but it would be good to reach out to this team to let them know that these changes are coming, so that they can update their problems to take advantage of this new styling and layout. 

At this point @clrux  you can do a final / second review (whichever it winds up being.) Thanks for flagging that course issue though!

marcotuts commented:
@clrux  - Addressed font-weight changes in commit.

clrux commented:
Thanks @marcotuts. Going to :+1: this one.

mekkz commented:
Jenkins test this please

mekkz commented:
Jenkins test this please

marcotuts commented:
Rebased from master to include OLI hinting work! (Task list above updated to reflect this.)  Sandbox includes course with testing for these new template, and I've verified the 5 main mobile inputs.

frrrances commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -224,7 +328,7 @@ div.problem {
 
       input {
-        border-color: green;
+        border-color: $correct;
```
nice!

frrrances commented on lms/static/sass/base/_variables.scss:
```diff
@@ -82,6 +82,28 @@ $gray-d2: shade($gray,40%); // #4c4c4c
 $gray-d3: shade($gray,60%); // #323232
 $gray-d4: shade($gray,80%); // #191919
 
+// TO-DO: once existing lms $blue is removed, change $cms-blue to $blue. 
```
Not sure if this is a holdover from an earlier PR, but I think it's better to clean up the existing blues instead of wiring in a whole new set, especially since we have made the base LMS blue a11y-friendly.

frrrances commented:
@marcotuts Overall, looking good. I made a few small code comments, but am assuming that it's best not to dig too deep into the xmodule stuff at this time. Two things I noticed though: 

1. The Single Page Demo looks messed up in several ways (spacing between inputs, label position in relation to the input, positioning of the grade mark, missing grade marks, and problem broken): 

![screen shot 2015-06-22 at 4 33 50 pm](https://cloud.githubusercontent.com/assets/4327102/8292445/b3c70ce4-18fd-11e5-80a2-b749f8a38974.png)

It may be that the broken problem caused them all but I figured I would bring your attention to it. 

2. The Hinting section was empty so I didn't proof any of those. 

Once you address my comments, :+1: - unless you need a review of the Hinting problems.

mekkz commented:
@frrrances Could you please copy/paste the text of that traceback?

frrrances commented:
@mekkz Here you go: 

```
Staff debug info: Traceback (most recent call last): File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/capa_base.py", line 1064, in check_problem correct_map = self.lcp.grade_answers(answers) File "/edx/app/edxapp/edx-platform/common/lib/capa/capa/capa_problem.py", line 378, in grade_answers return self._grade_answers(answers) File "/edx/app/edxapp/edx-platform/common/lib/capa/capa/capa_problem.py", line 440, in _grade_answers results = responder.evaluate_answers(self.student_answers, oldcmap) File "/edx/app/edxapp/edx-platform/common/lib/capa/capa/responsetypes.py", line 261, in evaluate_answers new_cmap = self.get_score(student_answers) File "/edx/app/edxapp/edx-platform/common/lib/capa/capa/responsetypes.py", line 1505, in get_score _(u"You may not use variables ({bad_variables}) in numerical problems.").format(bad_variables=undef_var.message) StudentInputError: You may not use variables (sdf) in numerical problems.
```

mekkz commented:
@frrrances @marcotuts That error message appears also when testing out the course on edge.edx.org, so that seems to be a separate bug (or feature). Do you think that's worth creating a bug ticket for?

marcotuts commented:
@mekkz feel free to great a separate ticket to improve validation for capa inputs. Having numerical input fail so spectacularly with texts hours be considered a bug. 

@frrances ! Thank you for catching those blues ! That must have been an earlier rebase issue I missed. I thought those has been cleaned up out of here earlier. 

I will look into the spacing and labeling issues you identified as well. My guess is that depending on the spacing I the common problem editor we output slightly different html. I'll see if we can clean up any of that, or if not create a separate story. Also, my apologies on the hinting I must not have published those In studio.

marcotuts commented:
I've addressed feedback from @frrrances on the PR, and also logged TNL bugs for the numerical input validation bug and the numerical / text input spacing issue that occurs sporadically in my testing. 

I've now published the hinting problems on the sandbox course and reverified those problems were working as expected.

mekkz commented:
@marcotuts Sounds good, I created an [issue](https://openedx.atlassian.net/browse/TNL-2558) on the TNL board.
